### PR TITLE
INTLY-6522 alerts for  cro resource state metrics

### PR DESCRIPTION
**Pull request by austincunningham**
# Description

Add alerts on the phase status returned by the cloud resource operator
https://issues.redhat.com/browse/INTLY-6522
follow on from https://issues.redhat.com/browse/INTLY-6521 where the metrics were exposed.

## Type of change

 
- [x] New feature (non-breaking change which adds functionality)

## Checklist
- [x] This change requires a documentation update  SOP url and doc need to be update , future jira for this work https://issues.redhat.com/browse/INTLY-7756
- [ ] Verified independently on a cluster by reviewer

### Verification
BYOC cluster required for verification
- clone this branch
- use this cro image quay.io/austincunningham/cloud-resource-operator:0.15.2 to expose the metrics needed
- install the latest version of the clould-resource-operator in rhmi
```bash
go get github.com/integr8ly/cloud-resource-operator@master
```
- `make cluster/prepare/project`
- `make cluster/prepare/crd`
- change `UseClusterStorage` to false in the rhmi cr
- deploy the cr `oc create -f deploy/crds/examples/rhmi.cr.yaml`
- `make code/run`
- check prometheus for new alerts
    - [productname]-PostgresResourceStatusPhaseFailed
    - [productname]-PostgresResourceStatusPhasePending
e.g.
image

    - [productname]-RedisResourceStatusPhaseFailed
    - [productname]-RedisResourceStatusPhasePending
image

### CODE REVIEW

File: pkg/resources/metrics.go [hunk 1]
```diff
@@ -78,6 +78,114 @@ func CreatePostgresConnectivityAlert(ctx context.Context, client k8sclient.Clien
...
+	postgresCRName := strings.Title(strings.Replace(cr.Name, "postgres-example-rhmi", "", -1))
+	alertName := postgresCRName + "PostgresResourceStatusPhaseFailed"
+	ruleName := fmt.Sprintf("resource-status-phase-failed-rule-%s", cr.Name)
+	alertExp := intstr.FromString(
+		fmt.Sprintf("%s{exported_namespace='%s',resourceID='%s',productName='%s',statusPhase='failed'} ",
+			croResources.DefaultPostgresStatusMetricName, cr.Namespace, cr.Name, productName),
+	)
+	alertDescription := fmt.Sprintf("The creation of the Postgres instance has been failing longer that %s. Postgres Custom Resource: %s in namespace %s (strategy: %s) for product: %s", alertFor, cr.Name, cr.Namespace, cr.Status.Strategy, productName)
+	labels := map[string]string{
+		"severity":    "critical",
```
Comment (david-martin) [inline pkg/resources/metrics.go:124]: @austincunningham  Just checking if there's a SOP lined up for new critical alerts

Comment (aidenkeating) [reply pkg/resources/metrics.go:124]: @david-martin That will be completed in https://issues.redhat.com/browse/INTLY-7756

Comment (david-martin) [reply pkg/resources/metrics.go:124]: Thanks @aidenkeating.
re: scheduling, let's aim to land the sop at the same time the rule rolls out in a release (if not already the plan)

Comment (austincunningham) [reply pkg/resources/metrics.go:124]: @david-martin It's not currently in the sprint, will see if we have capacity to add it .

File: pkg/resources/metrics.go [hunk 2]
```diff
@@ -78,6 +78,114 @@ func CreatePostgresConnectivityAlert(ctx context.Context, client k8sclient.Clien
...
+	if strings.ToLower(inst.Spec.UseClusterStorage) == "true" {
+		logrus.Info("skipping postgres state alert creation, useClusterStorage is true")
+		return nil, nil
+	}
+	productName := cr.Labels["productName"]
+	postgresCRName := strings.Title(strings.Replace(cr.Name, "postgres-example-rhmi", "", -1))
+	alertName := postgresCRName + "PostgresResourceStatusPhasePending"
+	ruleName := fmt.Sprintf("resource-status-phase-pending-rule-%s", cr.Name)
+	alertExp := intstr.FromString(
+		fmt.Sprintf("absent(%s{exported_namespace='%s',resourceID='%s',productName='%s',statusPhase='complete'} == 1)",
```
Comment (aidenkeating) [inline pkg/resources/metrics.go:92]: Now that we have the KSM-style alerts, could

```
statusPhase='in progress'} == 1
```

work here for pending checks?

File: pkg/resources/metrics.go [hunk 3]
```diff
@@ -78,6 +78,114 @@ func CreatePostgresConnectivityAlert(ctx context.Context, client k8sclient.Clien
...
+	if strings.ToLower(inst.Spec.UseClusterStorage) == "true" {
+		logrus.Info("skipping redis alert creation, useClusterStorage is true")
+		return nil, nil
+	}
+	productName := cr.Labels["productName"]
+	redisCRName := strings.Title(strings.Replace(cr.Name, "redis-example-rhmi", "", -1))
+	alertName := redisCRName + "RedisResourceStatusPhasePending"
+	ruleName := fmt.Sprintf("resource-status-phase-pending-rule-%s", cr.Name)
+	alertExp := intstr.FromString(
+		fmt.Sprintf("absent(%s{exported_namespace='%s',resourceID='%s',productName='%s',statusPhase='complete'} == 1)",
```
Comment (aidenkeating) [inline pkg/resources/metrics.go:146]: Same as above for this pending check

File: pkg/resources/metrics.go (hunk 2 above)
Comment (austincunningham) [reply pkg/resources/metrics.go:92]: Good question we have a number of phases, 
https://github.com/integr8ly/cloud-resource-operator/blob/master/pkg/apis/integreatly/v1alpha1/types/types.go#L8-L16

and just tracking `in progress` may leave us not tracking other phases. The absence of  `statusPhase='complete'` would be the only phase that is not pending.

File: pkg/resources/metrics.go (hunk 3 above)
Comment (austincunningham) [reply pkg/resources/metrics.go:146]: ^^

File: pkg/resources/metrics.go (hunk 2 above)
Comment (aidenkeating) [reply pkg/resources/metrics.go:92]: This makes sense to me, thanks

File: pkg/resources/metrics.go (hunk 1 above)
Comment (austincunningham) [reply pkg/resources/metrics.go:124]: @david-martin  added sop url here https://github.com/RHCloudServices/integreatly-help/pull/522

**Comment by aidenkeating** [comment]
changed the identifier of the user rhsso rds instance (this also changes the url). expected to see the connection fail and the cr status as failed. that worked

Screenshot 2020-05-26 at 12 20 11

changed the identifier back to normal (the url changes back to normal also). expected to see the alerts to back to green, that worked also.

Screenshot 2020-05-26 at 12 26 25

**Comment by aidenkeating** [comment]
@austincunningham Just need to set `OperatorVersionCloudResources` and `VersionCloudResources` in `rhmi_types.go` to `0.16.1`. It doesn't impact the install apart from setting metadata so I can test that in my update test

**Comment by matskiv** [comment]
I reviewed the changes in the manifests - all looks good 👍

**Comment by aidenkeating** [comment]
## Pre-upgrade
Screenshot 2020-05-26 at 16 25 24

## Post-upgrade

Screenshot 2020-05-26 at 16 39 21
Screenshot 2020-05-26 at 16 36 55
Screenshot 2020-05-26 at 16 31 17
Screenshot 2020-05-26 at 16 43 22

**Pull request merged**